rest-api/src/test/java/com/example/restapi/controllers/BookmarkControllerTest.java

```diff
@@ -0,0 +1,132 @@
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@WebMvcTest(BookmarkController.class)
+class BookmarkControllerTest {
+
+  @Autowired private MockMvc mockMvc;
+
+  @MockBean private BookmarkRepository bookmarkRepository;
+
+  @Test
+  void fetchBookmarks_shouldReturnListOfBookmarks() throws Exception {
+    Mockito.when(bookmarkRepository.findAll()).thenReturn(getSampleBookmarks());
+
+    mockMvc
+        .perform(MockMvcRequestBuilders.get("/api/v1/bookmarks"))
+        .andExpect(status().isOk())
+        .andExpect(jsonPath("$[0].id").value("1"))
+        .andExpect(jsonPath("$[1].id").value("2"));
+  }
+
+  @Test
+  void fetchBookmarkById_shouldReturnBookmark() throws Exception {
+    Mockito.when(bookmarkRepository.findById("1"))
+        .thenReturn(Optional.of(getSampleBookmarks().getFirst()));
+
+    mockMvc
+        .perform(MockMvcRequestBuilders.get("/api/v1/bookmarks/1"))
+        .andExpect(status().isOk())
+        .andExpect(jsonPath("$.id").value("1"));
+  }
+
+  @Test
+  void addBookmark_shouldCreateAndReturnBookmark() throws Exception {
+    Mockito.when(bookmarkRepository.save(Mockito.any(Bookmark.class)))
+        .thenReturn(getSampleBookmarks().getFirst());
+
+    mockMvc
+        .perform(
+            MockMvcRequestBuilders.post("/api/v1/bookmarks")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(
+                    "{\"id\":\"1\",\"title\":\"Title1\",\"description\":\"Description1\",\"url\":\"http://example.com/1\"}"))
+        .andExpect(status().isCreated())
+        .andExpect(jsonPath("$.id").value("1"));
+  }
+
+  @Test
+  void updateCompleteBookmark_shouldUpdateAndReturnBookmark() throws Exception {
+    Bookmark existingBookmark = getSampleBookmarks().getFirst();
+    Bookmark updatedBookmark =
+        new Bookmark(
+            "1",
+            "UpdatedTitle",
+            "UpdatedDescription",
+            "http://example.com/updated",
+            Instant.now(),
+            Instant.now());
+
+    Mockito.when(bookmarkRepository.findById("1")).thenReturn(Optional.of(existingBookmark));
+    Mockito.when(bookmarkRepository.save(Mockito.any(Bookmark.class))).thenReturn(updatedBookmark);
+
+    mockMvc
+        .perform(
+            MockMvcRequestBuilders.put("/api/v1/bookmarks/1")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(
+                    "{\"id\":\"1\",\"title\":\"UpdatedTitle\",\"description\":\"UpdatedDescription\",\"url\":\"http://example.com/updated\"}"))
```

Comment:
json would be way more readable if we would use text block here like
```
String content = """
{
  "someProperty": "someValue"
}
""";
```

Reply:
sure, I'll update it.